### PR TITLE
Refactoring of Consensus.joinDuringCreation

### LINKED ISSUE #83: Refactor Consensus::JoinDuringCreation
- Update the params of Consensus::joinDuringCreation
  - joinDuringCreation(bytes32 _metachainId, address _withdrawalAddress).
- Add a require/modifier that the value of CoreLifetime should be Creation.
- Remove validateJoinParams function.
- Remove isCore require.
- Call reputation.registerjoin.
- Call core.joinDuringCreation(msg.sender).
- if the core.joinDuringCreation(msg.sender) returns true, then call ConsensusGateway.declareKernelOpened(core)

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below
- Updated interface for method `consensus.joinDuringCreation`. Core address is removed from `joinDuringCreation` method.
```
function joinDuringCreation(
        bytes32 _metachainId,
        address _withdrawalAddress
    )
```

- Updated interface for method `consensus.join`. Core address is removed from `join` method.
```
function join(
        bytes32 _metachainId,
        address _withdrawalAddress
    )
```

- Implemented ConsensusGateway interface since `ConsensusGateway::declareOpenKernel` is called when joined validator count is greater than minimum validator count.
Specification reference: https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20191210-consensusgateway-tasks.md

- Discussion on feeGasPrice and feeGasLimit is here:
  https://github.com/mosaicdao/mosaic-1/issues/115 

Fixes #83

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -116,6 +117,12 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
     /** Coinbase split per mille */
     uint256 public coinbaseSplitPerMille;
 
+    /** Gas price to calculate reward */
+    uint256 public feeGasPrice;
```
Comment (0xsarvesh) [inline contracts/consensus/Consensus.sol:145]: How this value can be set? 
I can't find initialization in the setup? should it be constant for now?

Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol:145]: Same comment for feeGasLimit

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -505,63 +515,67 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
         address _withdrawalAddress
     )
         external
     {
-        // Validate the join params.
-        validateJoinParams(_metachainId, _core, _withdrawalAddress);
+        address core = assignments[_metachainId];
+        require(
+            core != address(0),
+            "Core address is 0."
```
Comment (0xsarvesh) [inline contracts/consensus/Consensus.sol]: Core doesn't exist for given metachain ?

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -505,63 +515,67 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
```
Comment (0xsarvesh) [inline contracts/consensus/Consensus.sol]: Same suggestioh.

File: contracts/consensus/Consensus.sol [hunk 4]
```diff
@@ -570,10 +584,16 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
 
         // Join in core contract.
         (uint256 validatorCount, uint256 minValidatorCount) =
-            CoreI(_core).joinDuringCreation(msg.sender);
+            CoreI(core).joinDuringCreation(msg.sender);
 
         if (validatorCount >= minValidatorCount) {
-            coreLifetimes[_core] = CoreLifetime.genesis;
+            coreLifetimes[core] = CoreLifetime.genesis;
+            ConsensusGatewayI consensusGateway = consensusGateways[_metachainId];
```
Comment (0xsarvesh) [inline contracts/consensus/Consensus.sol:617]: it should check consensusGateway exists for metachain Id?

File: test/consensus/join.js [hunk 5]
```diff
@@ -55,57 +54,20 @@ contract('Consensus::join', (accounts) => {
...
-
-    it('should fail when core address is 0', async () => {
-      const params = Object.assign(
-        {},
-        joinParams,
-        { core: Utils.NULL_ADDRESS },
-      );
-
+    it('should fail when metachainId is invalid', async () => {
+      const joinParams1 = {
```
Comment (0xsarvesh) [inline test/consensus/join.js]: This variable name does not tell me.

**Review by 0xsarvesh — DISMISSED** [review]
Looks good. 
Some feedback and questions.

File: contracts/consensus/Consensus.sol (hunk 1 above)
Comment (abhayks1) [reply contracts/consensus/Consensus.sol:145]: These two parameters will be provided and managed by the techgov. The logic regarding updating these values will later be finalized in either setup function or some other way.
In current implementation initialization is done in setup function.

Detailed discussion on the ticket: https://github.com/mosaicdao/mosaic-1/issues/115

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (abhayks1) [reply contracts/consensus/Consensus.sol]: Done!

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (abhayks1) [reply contracts/consensus/Consensus.sol]: Done!

File: contracts/consensus/Consensus.sol (hunk 4 above)
Comment (abhayks1) [reply contracts/consensus/Consensus.sol:617]: Added require.

File: test/consensus/join.js (hunk 5 above)
Comment (abhayks1) [reply test/consensus/join.js]: Updated!

**Comment by abhayks1** [comment]
PR is ready for review again.

File: contracts/consensus/Consensus.sol [hunk 6]
```diff
@@ -598,10 +611,21 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+            require(
+                address(consensusGateway) != address(0),
+                "Consensus gateway does not exist for given metachain."
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: Can `consensusGateway` address be `0`? 
This should never happen.
Either remove this or make this assert.

File: contracts/test/consensus_gateway/SpyConsensusGateway.sol [hunk 7]
```diff
@@ -15,10 +15,28 @@ pragma solidity ^0.5.0;
...
 
-contract SpyConsensusGateway is MasterCopyNonUpgradable {
+contract SpyConsensusGateway is MasterCopyNonUpgradable, ConsensusGatewayI {
+
+    address public spyCore;
+    uint256 public spyFeeGasPrice;
+    uint256 public spyFeeGasLimit;
 
     function setup() public {
         // do nothing
```
Comment (deepesh-kn) [inline contracts/test/consensus_gateway/SpyConsensusGateway.sol:27]: If this is not being used in any tests, add assert here to break.

File: test/consensus/join_during_creation.js [hunk 8]
```diff
@@ -52,57 +54,19 @@ contract('Consensus::joinDuringCreation', (accounts) => {
...
 
   contract('Negative Tests', async () => {
-    it('should fail when chain id is 0', async () => {
-      const params = Object.assign(
+    it('should fail when metachainId is invalid', async () => {
+      const invalidJoinParams = Object.assign(
         {},
         joinParams,
-        { metachainId: Utils.NULL_ADDRESS },
+        { metachainId: '0x0000000000000000000000000000000000000000' },
```
Comment (deepesh-kn) [inline test/consensus/join_during_creation.js]: use Utils.ZERO_BYTES32

File: test/consensus/join_during_creation.js [hunk 9]
```diff
@@ -161,17 +121,20 @@ contract('Consensus::joinDuringCreation', (accounts) => {
...
         joinParams.txOptions.from,
         'Validator address not set in spy core contract',
       );
+
+      const spyCoreFromConsensusGateway = await consensusGateway.spyCore.call();
+      assert.strictEqual(
+        spyCoreFromConsensusGateway,
+        core.address,
+        'Core address not set in spy consensus gateway contract',
+      );
```
Comment (deepesh-kn) [inline test/consensus/join_during_creation.js:138]: Please assert that all the spy params are called. Or add a todo/task note.

Comment (abhayks1) [reply test/consensus/join_during_creation.js:138]: Done!

File: test/consensus/join_during_creation.js (hunk 8 above)
Comment (abhayks1) [reply test/consensus/join_during_creation.js]: Done!

File: contracts/test/consensus_gateway/SpyConsensusGateway.sol (hunk 7 above)
Comment (abhayks1) [reply contracts/test/consensus_gateway/SpyConsensusGateway.sol:27]: It's being used when Axiom.deployMetachainProxies is called.
`ConsensusGateway` setup data is passed as argument.

File: contracts/consensus/Consensus.sol (hunk 6 above)
Comment (abhayks1) [reply contracts/consensus/Consensus.sol]: Updated with assert.

**Comment by abhayks1** [comment]
PR is available for review

**Review by deepesh-kn — APPROVED** [review]
LGTM

**Pull request merged**